### PR TITLE
RGD with matching on tautomers of core

### LINKED ISSUE #6609: RGD to support tautomers of core
**Is your feature request related to a problem? Please describe.**

In RGD it would be nice if tautomers of a core could also match input structures.  For example, these cores:

image

Would each match both of these structures:

image

Currently, each core matches only one structure.

**Describe the solution you'd like**

Would like to integrate the `TautomerQuery` matcher into RGD's `RGroupDecomp` class.

**Describe alternatives you've considered**

Alternatively, the core could be enumerated and all its tautomers used as the input cores for the decomposition.  This would be complex for the user and would not allow RGD to take advantage of the optimized tautomer matching.

**Additional context**

I plan to work on a PR to resolve.

### PR DESCRIPTION
**Pull request by jones-gareth**
#### Reference Issue

Fixes #6609.

#### What does this implement/fix? Explain your changes.

I added an `allowTautomerCore` parameter to  `RGroupDecompositionParameters`.  When set RGD will match the core to the 
input structure using a `TautomerQuery` object, if possible.

The different tautomer forms in #6609 now decompose as desired:

image

I've included C++, Python and C# tests.

#### Any other comments?

Also added enum support to the SWIG interface file for RGroupDecomposition.

### CODE REVIEW

**Comment by bp-kelley** [comment]
Is there any reason this shouldn't be the default?

File: Code/GraphMol/RGroupDecomposition/RGroupCore.cpp
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupCore.cpp:826]: I think this could just be 

```
RWMol copy(*matchingMol);
```

and not a pointer.

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:135]: if tautomerQuery is nullptr, no matches should be performed,  shouldn't we call back to the original query?

File: Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h:78]: Consider making this the default (i.e. allowTautomerCore = true)

**Review by bp-kelley — CHANGES_REQUESTED** [review]

File: Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h
Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h:78]: Works for me, but given that it will change the behavior RGD it would be best if you and/or Greg make the call,

File: Code/GraphMol/RGroupDecomposition/RGroupCore.cpp
Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupCore.cpp]: I've made that change

File: Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupDecompParams.h:78]: I'm very wary of changing default behavior unless the defaults produce confusing results.
Given that most RDKit code does not behave in a tautomer-insensitive way, I'd rather not make it the default here.

If it turns out to create problems, we can change the default in a later release.

**Comment by jones-gareth** [comment]
> @jones-gareth I like the idea of this one and think it would be a nice addition. I wonder though if it wouldn't make sense to just go ahead and add generalized substructure support here via the extended query mol interface (added in #6443) It seems like there could be a use case for supporting cores with things like link nodes: image or variable attachment points (though those might be harder to handle).
> 
> You're closer to actual users of this, so I defer to you as to whether or not it's worth doing.

The generalized query support is really nice.  I will take a look and see if I can substitute the generalized query for the tautomer uery.  Looks like it should be straightforward.

**Comment by bp-kelley** [comment]
I would very much appreciate the generalized querying as well.

**Comment by jones-gareth** [comment]
I've added a test for enumeration of cores

File: Code/GraphMol/RGroupDecomposition/RGroupDecompData.cpp
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupDecompData.cpp:31]: That was almost too easy to support. :-)

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:135]: it doesn't look to me like `useNormalMatch` is actually doing anything here.
The `else` on line 138 is only evaluated if `params().doTautomers` is false, but `useNormalMatch` only gets set to false if `params().doTautomers` is true.

Am I missing something obvious or do you need to get rid of the `else` on l138?

**Review by greglandrum — COMMENTED** [review]
I'm also confused by the new logic in RGroupDecomp::add

File: Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp
Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp:135]: Clearly I write very confusing code, but I do believe that this logic is necessary.

`useNormalMatch` gets set to false if `params().doTautomers` is true *and* `tautomerQuery != nullptr`.  `useNormalMatch` is needed for the cases where `params().doTautomers` is true *and* `tautomerQuery == nullptr`- which will happen if the core cannot be Kekulized.

**Comment by jones-gareth** [comment]
> I'm also confused by the new logic in RGroupDecomp::add

Github does not seem to be threading the replies correctly - this is a copy of my comment above.

--
Clearly I write very confusing code, but I do believe that this logic is necessary.

`useNormalMatch` gets set to false if `params().doTautomers` is true *and* `tautomerQuery != nullptr`. `useNormalMatch` is needed for the cases where `params().doTautomers` is true *and* `tautomerQuery == nullptr`- which will happen if the core cannot be Kekulized.

**Comment by jones-gareth** [comment]
I have admitted defeat and removed `useNormalMatch` from RGroupDecomp::add

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
@bp-kelley : are you happy with the changes here?

File: Code/GraphMol/RGroupDecomposition/RGroupCore.cpp
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupCore.cpp:826]: I'm assuming the reason for the copy is because we need to call updatePropertyCache?

File: Code/GraphMol/RGroupDecomposition/RGroupCore.cpp
Comment (bp-kelley) [inline Code/GraphMol/RGroupDecomposition/RGroupCore.cpp:821]: Could we also indicate here that when called, if a tautomerQuery is returned, then this will automatically use it and if not the normal match will be performed.   This will document the confusion I had below.

Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupCore.cpp:821]: I just have a document suggestion, otherwise LGTM!  Can't wait to use it.

Comment (jones-gareth) [reply Code/GraphMol/RGroupDecomposition/RGroupCore.cpp:821]: I've added code documentation round Brian's issue.

**Review by greglandrum — APPROVED** [review]
LGTM

**Comment by greglandrum** [comment]
Will merge once the CI builds finish (some of them failed due to network problems)

**Pull request merged**
